crates/tower-cmd/src/run.rs

```diff
@@ -216,86 +256,146 @@ async fn do_run_remote(
                 output::write(&link_line);
                 output::newline();
             }
+            Ok(())
         }
     }
 }
 
+trait OutputSink: Send + Sync {
+    fn send_line(&self, line: String);
+    fn send_error(&self, error: String);
+}
+
+struct StdoutSink;
+impl OutputSink for StdoutSink {
+    fn send_line(&self, line: String) {
+        output::write(&line);
+    }
+    fn send_error(&self, error: String) {
+        output::failure(&error);
+    }
+}
+
+struct ChannelSink(tokio::sync::mpsc::UnboundedSender<String>);
+impl OutputSink for ChannelSink {
+    fn send_line(&self, line: String) {
+        let _ = self.0.send(line);
+    }
+    fn send_error(&self, error: String) {
+        let _ = self.0.send(error);
+    }
+}
+
+pub async fn do_follow_run_capture(
+    config: Config,
+    run: &Run,
+) -> Result<Vec<String>> {
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+    let sink = ChannelSink(tx);
+    
+    let result = do_follow_run_impl(config, run, &sink, false).await;
+    
+    let mut output_lines = Vec::new();
+    while let Ok(line) = rx.try_recv() {
+        output_lines.push(line);
+    }
+    
+    result.map(|_| output_lines)
+}
+
 async fn do_follow_run(
     config: Config,
     run: &Run,
 ) {
-    let mut spinner = output::spinner("Waiting for run to start...");
+    let sink = StdoutSink;
+    let _ = do_follow_run_impl(config, run, &sink, true).await;
+}
 
+async fn do_follow_run_impl(
+    config: Config,
+    run: &Run,
+    sink: &dyn OutputSink,
+    enable_ctrl_c: bool,
+) -> Result<()> {
     match wait_for_run_start(&config, &run).await {
         Err(err) => {
-            spinner.failure();
```

Comment:
we really don't want this in the main output, because then it would generate a thousand spinning braille events that would be parsed as separate lines in the mcp server.

Note to self: I need to double check it's still there when running directly in the cli 🙃